File: src/main/java/net/azzy/pulseflux/blockentity/logistic/transport/EverfullUrnEntity.java
package net.azzy.pulseflux.blockentity.logistic.transport;

import io.github.cottonmc.cotton.gui.PropertyDelegateHolder;
import net.azzy.pulseflux.mixin.BucketFluidAccessor;
import net.azzy.pulseflux.util.fluid.FluidHelper;
import net.azzy.pulseflux.util.fluid.FluidHolder;
import net.azzy.pulseflux.util.fluid.FluidPackage;
import net.azzy.pulseflux.util.gui.ExtendedPropertyDelegate;
import net.azzy.pulseflux.util.interaction.HeatTransferHelper;
import net.azzy.pulseflux.util.interaction.InventoryWrapper;
import net.azzy.pulseflux.util.networking.Syncable;
import net.fabricmc.fabric.api.block.entity.BlockEntityClientSerializable;
import net.minecraft.block.BlockState;
import net.minecraft.block.entity.BlockEntity;
import net.minecraft.fluid.Fluid;
import net.minecraft.fluid.Fluids;
import net.minecraft.inventory.Inventories;
import net.minecraft.item.BucketItem;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.CompoundTag;
import net.minecraft.screen.PropertyDelegate;
import net.minecraft.util.Tickable;
import net.minecraft.util.collection.DefaultedList;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;

import static net.azzy.pulseflux.registry.BlockEntityRegistry.EVERFULL_URN_ENTITY;

public class EverfullUrnEntity extends BlockEntity implements BlockEntityClientSerializable, Syncable, FluidHolder, InventoryWrapper, PropertyDelegateHolder, Tickable {

    private long pressure;
    private FluidPackage tank = FluidHelper.empty();
    private DefaultedList<ItemStack> inventory = DefaultedList.ofSize(1, ItemStack.EMPTY);

    public EverfullUrnEntity() {
        super(EVERFULL_URN_ENTITY);
    }

    @Override
    public void tick() {
        for(Direction direction : Direction.values())
            forcePush(world, pos.offset(direction), direction, 1, 1000);
    }

    @Override
    public CompoundTag toTag(CompoundTag tag) {
        Inventories.toTag(tag, inventory);
        tag.putLong("pressure", pressure);
        tag.put("tank", tank.toTag());
        return super.toTag(tag);
    }

    @Override
    public void fromTag(BlockState state, CompoundTag tag) {
        Inventories.fromTag(tag, inventory);
        pressure = tag.getLong("pressure");
        tank = FluidPackage.fromTag(tag.getCompound("tank"));
        super.fromTag(state, tag);
    }

    @Override
    public void fromClientTag(CompoundTag compoundTag) {
        pressure = compoundTag.getLong("pressure");
        tank = FluidPackage.fromTag(compoundTag.getCompound("tank"));
    }

    @Override
    public CompoundTag toClientTag(CompoundTag compoundTag) {
        compoundTag.putLong("pressure", pressure);
        compoundTag.put("tank", tank.toTag());
        return compoundTag;
    }

    @Override
    public void synchronize(SyncPacket packet) {
        this.pressure = (long) packet.unpack();
        if(!inventory.isEmpty() && inventory.get(0).getItem() instanceof BucketItem){
            Fluid fluid = ((BucketFluidAccessor) inventory.get(0).getItem()).getFluid();
            tank = fluid == Fluids.EMPTY ? FluidHelper.empty() : new FluidPackage(fluid, HeatTransferHelper.translateBiomeHeat(world.getBiome(pos)), pressure, 8000, false);
        }
        else
            tank = FluidHelper.empty();
        markDirty();
        sync();
    }

    @Override
    public FluidPackage getFluid() {
        return tank;
    }

    @Override
    public void setFluid(FluidPackage fluid) {
    }

    @Override
    public boolean testFluid(FluidPackage fluid) {
        return true;
    }

    @Override
    public void recalcPressure() {

    }

    @Override
    public long addFluid(long amount) {
        return 0;
    }

    @Override
    public long extractFluid(long amount) {
        return 0;
    }

    @Override
    public boolean gasCarrying() {
        return false;
    }

    @Override
    public DefaultedList<ItemStack> getItems() {
        return inventory;
    }

    private final PropertyDelegate delegate = new ExtendedPropertyDelegate() {
        @Override
        public long getLong(int index) {
            return pressure;
        }

        @Override
        public void setLong(int index, long value) {

        }

        @Override
        public BlockPos getpos() {
            return pos;
        }
    };

    @Override
    public PropertyDelegate getPropertyDelegate() {
        return delegate;
    }
}

File: src/main/java/net/azzy/pulseflux/blockentity/logistic/transport/EverfullUrnEntity.kt
package net.azzy.pulseflux.blockentity.logistic.transport

import dev.technici4n.fasttransferlib.api.Simulation
import dev.technici4n.fasttransferlib.api.fluid.FluidIo
import io.github.cottonmc.cotton.gui.PropertyDelegateHolder
import net.azzy.pulseflux.client.util.RenderMathHelper
import net.azzy.pulseflux.mixin.BucketFluidAccessor
import net.azzy.pulseflux.registry.BlockEntityRegistry.EVERFULL_URN_ENTITY
import net.azzy.pulseflux.util.gui.ExtendedPropertyDelegate
import net.azzy.pulseflux.util.interaction.HeatTransferHelper
import net.azzy.pulseflux.util.interaction.InventoryWrapper
import net.azzy.pulseflux.util.interaction.PressureHolder
import net.azzy.pulseflux.util.networking.Syncable
import net.azzy.pulseflux.util.networking.Syncable.SyncPacket
import net.fabricmc.fabric.api.block.entity.BlockEntityClientSerializable
import net.minecraft.block.BlockState
import net.minecraft.block.entity.BlockEntity
import net.minecraft.fluid.Fluid
import net.minecraft.fluid.Fluids
import net.minecraft.inventory.Inventories
import net.minecraft.item.BucketItem
import net.minecraft.item.ItemStack
import net.minecraft.nbt.CompoundTag
import net.minecraft.screen.PropertyDelegate
import net.minecraft.util.Tickable
import net.minecraft.util.collection.DefaultedList
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction

class EverfullUrnEntity : BlockEntity(EVERFULL_URN_ENTITY), BlockEntityClientSerializable, PressureHolder, Syncable, InventoryWrapper, PropertyDelegateHolder, Tickable, FluidIo {

    private var pressure: Long = 0
    var fluid: Fluid = Fluids.EMPTY
    private val inventory = DefaultedList.ofSize(1, ItemStack.EMPTY)

    override fun tick() {
        //for (direction in Direction.values())
    }

    override fun toTag(tag: CompoundTag): CompoundTag {
        Inventories.toTag(tag, inventory)
        tag.putLong("pressure", pressure)
        return super.toTag(tag)
    }

    override fun fromTag(state: BlockState, tag: CompoundTag) {
        Inventories.fromTag(tag, inventory)
        fluid = (inventory[0].item as BucketFluidAccessor).fluid
        pressure = tag.getLong("pressure")
        super.fromTag(state, tag)
    }

    override fun fromClientTag(compoundTag: CompoundTag) {
        pressure = compoundTag.getLong("pressure")
        Inventories.fromTag(compoundTag, inventory)
        fluid = (inventory[0].item as BucketFluidAccessor).fluid
    }

    override fun toClientTag(compoundTag: CompoundTag): CompoundTag {
        Inventories.toTag(compoundTag, inventory)
        compoundTag.putLong("pressure", pressure)
        return compoundTag
    }

    override fun synchronize(packet: SyncPacket) {
        pressure = packet.unpack() as Long
        fluid = if (!inventory.isEmpty() && inventory[0].item is BucketItem) {
            (inventory[0].item as BucketFluidAccessor).fluid
        } else Fluids.EMPTY
        markDirty()
        sync()
    }

    override fun getItems(): DefaultedList<ItemStack> {
        return inventory
    }

    private val delegate: PropertyDelegate = object : ExtendedPropertyDelegate {
        override fun getLong(index: Int): Long {
            return pressure
        }

        override fun setLong(index: Int, value: Long) {}
        override fun getpos(): BlockPos {
            return pos
        }
    }

    override fun markDirty() {
        super<BlockEntity>.markDirty()
    }

    override fun getPropertyDelegate(): PropertyDelegate {
        return delegate
    }

    override fun getFluidSlotCount(): Int {
        return 1
    }

    override fun getFluid(p0: Int): Fluid {
        return fluid
    }

    override fun getFluidAmount(p0: Int): Long {
        return Long.MAX_VALUE
    }

    override fun getPressure(): Long {
        return pressure
    }

    override fun setPressure(pressure: Long) {
        this.pressure = pressure
    }

    override fun supportsFluidExtraction(): Boolean {
        return true
    }

    override fun supportsFluidInsertion(): Boolean {
        return fluid == Fluids.EMPTY
    }

    override fun extract(fluid: Fluid?, maxAmount: Long, simulation: Simulation?): Long {
        return maxAmount
    }

    override fun insert(fluid: Fluid?, amount: Long, simulation: Simulation?): Long {
        return 0L
    }
}